src/ccl_background.c

```diff
@@ -0,0 +1,1208 @@
+
+  if (chistatus) {
+    free(a);
+    free(y);
+    gsl_spline_free(E);
+    gsl_spline_free(chi);
+    gsl_spline_free(achi);
+    *status = CCL_ERROR_SPLINE;
+    strcpy(cosmo->status_message,"ccl_background.c: ccl_cosmology_compute_distances(): Error creating  a(chi) spline\n"); 
+    return;
+  }
+
+  if(cosmo->data.accelerator==NULL)
+    cosmo->data.accelerator=gsl_interp_accel_alloc();
+  cosmo->data.E = E;
+  cosmo->data.chi = chi;
+  cosmo->data.achi=achi;
+  cosmo->computed_distances = true;
+    
+  free(a);
+  free(y);
+
+}
+
+
+/* ----- ROUTINE: ccl_cosmology_compute_growth ------
+INPUT: cosmology
+TASK: if not already there, make a table of growth function and growth rate
+      normalize growth to input parameter growth0
+*/
+
+void ccl_cosmology_compute_growth(ccl_cosmology * cosmo, int * status)
+{
+	
+  // This is not valid for massive neutrinos; if we have massive neutrinos, exit.
+  if (cosmo->params.N_nu_mass>0){
+	  *status = CCL_ERROR_NOT_IMPLEMENTED;
+	  strcpy(cosmo->status_message,"ccl_background.c: ccl_cosmology_compute_growth(): Support for the growth rate in cosmologies with massive neutrinos is not yet implemented.\n");
+	  return; 
+  }
+	
+  if(cosmo->computed_growth)
+    return;
+
+  // Create logarithmically and then linearly-spaced values of the scale factor
+  int  chistatus = 0, na = ccl_splines->A_SPLINE_NA+ccl_splines->A_SPLINE_NLOG-1;
+  double * a = ccl_linlog_spacing(ccl_splines->A_SPLINE_MINLOG, ccl_splines->A_SPLINE_MIN, ccl_splines->A_SPLINE_MAX, ccl_splines->A_SPLINE_NLOG, ccl_splines->A_SPLINE_NA);
+  if (a==NULL || 
+      (fabs(a[0]-ccl_splines->A_SPLINE_MINLOG)>1e-5) || 
+      (fabs(a[na-1]-ccl_splines->A_SPLINE_MAX)>1e-5) || 
+      (a[na-1]>1.0)
+      ) {
+    *status = CCL_ERROR_LINSPACE;
+    strcpy(cosmo->status_message,"ccl_background.c: ccl_cosmology_compute_growth(): Error creating logarithmically and then linear spacing in a\n");
+    return;
+  }
+
+  gsl_integration_cquad_workspace * workspace=NULL;
+  gsl_function F;
+  gsl_spline *df_a_spline=NULL;
+  if(cosmo->params.has_mgrowth) {
+    double *df_arr=malloc(na*sizeof(double));
+    if(df_arr==NULL) {
+      free(a);
+      *status=CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,"ccl_background.c: ccl_cosmology_compute_distances(): ran out of memory\n");
+      return;
+    }
+    //Generate spline for Delta f(z) that we will then interpolate into an array of a
+    gsl_spline *df_z_spline=gsl_spline_alloc(A_SPLINE_TYPE,cosmo->params.nz_mgrowth);
+    chistatus=gsl_spline_init(df_z_spline,cosmo->params.z_mgrowth,cosmo->params.df_mgrowth,
+			      cosmo->params.nz_mgrowth);
+    
+    if(chistatus) {
+      free(a);
+      free(df_arr);
+      gsl_spline_free(df_z_spline);
+      *status = CCL_ERROR_SPLINE;
+      strcpy(cosmo->status_message,"ccl_background.c: ccl_cosmology_compute_growth(): Error creating Delta f(z) spline\n");
+      return;
+    }
+    for (int i=0; i<na; i++) {
+      if(a[i]>0) {
```

Comment:
We don't control the global minimum A value for splines set in `ccl_splines->A_SPLINE_MIN` since people can pass in any value via the ini file. Thus I think we should keep this check. I agree though passing less than zero is silly. Someone might pass zero though I suppose.